packages/grainIntegration-merkle/contracts/MerkleRedeem.sol

```diff
@@ -0,0 +1,193 @@
+
+    // Recorded weeks
+    mapping(uint => bytes32) public weekMerkleRoots;
+    mapping(uint => mapping(address => bool)) public claimed;
+    mapping(uint => uint64) public claimTimes;
+
+    constructor(
+        address _token
+    ) public {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        token = IERC20(_token);
+    }
+
+    function disburse(
+        address _liquidityProvider,
+        uint _balance
+    )
+        private
+    {
+        if (_balance > 0) {
+            emit Claimed(_liquidityProvider, _balance);
+            require(token.transfer(_liquidityProvider, _balance), "ERR_TRANSFER_FAILED");
+        }
+    }
+
+    /**
+      * @dev  ensure the contract's token balance is sufficient before calling
+      *       a claimWeek function.
+      */
+    function claimWeek(
+        address _liquidityProvider,
+        uint _week,
+        uint _claimedBalance,
+        bytes32[] memory _merkleProof
+    )
+        public
+    {
+        require(!claimed[_week][_liquidityProvider]);
+        require(block.timestamp > claimTimes[_week], "Distribution Still Paused");
+        require(verifyClaim(_liquidityProvider, _week, _claimedBalance, _merkleProof), 'Incorrect merkle proof');
+
+        claimed[_week][_liquidityProvider] = true;
+        disburse(_liquidityProvider, _claimedBalance);
+    }
+
+    struct Claim {
+        uint week;
+        uint balance;
+        bytes32[] merkleProof;
+    }
+
+    /**
+      * @dev  ensure the contract's token balance is sufficient before calling
+      *       a claimWeeks function.
+      */
+    function claimWeeks(
+        address _liquidityProvider,
+        Claim[] memory claims
+    )
+        public
+    {
+        uint totalBalance = 0;
+        Claim memory claim ;
+        for(uint i = 0; i < claims.length; i++) {
+            claim = claims[i];
+
+            require(block.timestamp > claimTimes[claim.week], "Distribution Still Paused");
+            require(!claimed[claim.week][_liquidityProvider]);
+            require(verifyClaim(_liquidityProvider, claim.week, claim.balance, claim.merkleProof), 'Incorrect merkle proof');
+
+            totalBalance += claim.balance;
+            claimed[claim.week][_liquidityProvider] = true;
+        }
+        disburse(_liquidityProvider, totalBalance);
+    }
+
+    /**
+     * Get the claim status for a user for a slice of distributions.
+     * This will error if an index for non-existent distribution is
+     * passed.
+     */
+    function claimStatus(
+        address _contributor,
+        uint _begin,
+        uint _end
+    )
+        external
+        view
+        returns (bool[] memory)
+    {
+        uint size = 1 + _end - _begin;
+        bool[] memory arr = new bool[](size);
+        for(uint i = 0; i < size; i++) {
+            arr[i] = claimed[_begin + i][_contributor];
+        }
+        return arr;
+    }
+
+    /**
+     * Get the bytes32 hash merkle root for a slice of all valid
```

Comment:
For these slice methods, it'd be nice to include in the docstring whether begin and end are inclusive or exclusive.

Reply:
done